Redux Saga

PranavBakre opened the pull request:
set jsconfig for simpler paths
redux saga and redux logger middleware added

PranavBakre committed:
redux and a demo page
set jsconfig for simpler paths
redux saga and redux logger middleware added

PranavBakre committed:
useless constructor

PranavBakre committed:
Hashrouter

AniketRaj314 commented:
Redux part is self-explanatory. Saga is the one I need to focus on.

PranavBakre commented:
> Redux part is self-explanatory. Saga is the one I need to focus on.

Check the DemoSaga generator and logger output on console

PranavBakre committed:
Added Reducer for Login Actions

Login => initiates the login action which is picked by the saga
middleware to perform api call to access_token and receive token.

LoginSuccess => Called when Login is Successful, adds access token to
the store and sets the isLoggedIn to true

LoginFailure => Called on failure and dispatches error message to the
store to be displayed in the toast

moved Login api call from the login page to the LoginSaga to be called
on Login action dispatch

Removed demo reducer and its related pages and files

Added redux-persist package to the application. To be implemented soon
after the logout reducer is added (most probably as a root reducer to
clear the entire store on logout)

TODO:
1. UserReducer and UserSaga => shifting getUser from Login page into UserSaga
2. Adding RefreshToken and TokenExpiry into Login
3. Configuring Redux-persist for maintaining state in local storage
4. Removing react-cookie from the application

akhilnarang committed:
frontend: format files after redux bringup

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ksdfg committed:
use redux-saga to fetch user details on login

Signed-off-by: ksdfg <ksdfg123@gmail.com>

akhilnarang commented on src/pages/login/login.page.jsx:
```diff
@@ -27,13 +28,18 @@ class Login extends Component {
+            this.props.FetchUserMe();
+        }
+
+        if (prevProps.user !== this.props.user && this.props.user) {
```
Just curious, shouldn't you first check `this.props.user` then check for inequality? Or any specific reason for this

PranavBakre replied:
no specific reason as such

ksdfg replied:
well, the first part is to check if the property has actually changed since the last time component was updated (basically so that we don't trigger this function twice for same user when some other property has changed) and the second part is to check that the current value of the property is not set to `null`, since that is the default value when no current user details are fetched.

ksdfg replied:
doing it in this order made sense since the second part doesn't need to be checked if the property hasn't changed at all

akhilnarang replied:
Ah, I was thinking from the point of null check first

PranavBakre replied:
I was considering from the point where both the conditions were a deciding factor.. in that case order shouldnt matter as much so my reply

ksdfg replied:
Yeah, in the end the order doesn't matter since both are equally important, just made sense to me sequence wise that way.

ksdfg committed:
rename userDetails to currentUser to be more accurate as well as avoid confusion later

Signed-off-by: ksdfg <ksdfg123@gmail.com>

PranavBakre committed:
Added RootReducer and RootSaga

Renamed LoginReducer and LoginSaga to AuthReducer and AuthSaga to
signify the presence of LogOut action

Implemented RootReducer to allow for Logout and decoupled RootReducer
and RootSaga from store.js

Added a logout button to /login for easy cleanup of store during testing
 Changes to be committed:
	modified:   src/pages/login/login.page.jsx
	renamed:    src/redux/login/action.js -> src/redux/auth/action.js
	renamed:    src/redux/login/action.types.js -> src/redux/auth/action.types.js
	renamed:    src/redux/login/reducer.js -> src/redux/auth/reducer.js
	renamed:    src/redux/login/saga.js -> src/redux/auth/saga.js
	new file:   src/redux/reducer.js
	new file:   src/redux/saga.js
	modified:   src/redux/store.js
	modified:   src/redux/user/saga.js

akhilnarang committed:
format

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

PranavBakre committed:
Added FetchUser in userReducer
As per Akhil's Request, Logout button will be invisible till login takes place. Not done for Submit

Also wrote fetch two styles. using yield call and yield. need to identify which is better and should be preferred.  Commented yield call fetch

PranavBakre committed:
Merge Conflict resolution

PranavBakre committed:
commented call import to let the build run successfully.

PranavBakre committed:
created http-client service.
(basically moved fetch into one file so that some code wont be repeated?
Updated saga.js files to use http-client instead of fetch and use yield call instead of yield

akhilnarang committed:
frontend: some cleanup and formatting

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ksdfg committed:
redux/auth/saga.js: reformatting

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
redux/user/saga.js: use httpClient service instead of fetch in FetchUser

Signed-off-by: ksdfg <ksdfg123@gmail.com>

PranavBakre committed:
removed unnecessary braces

AniketRaj314 reviewed (approved):
LGTM

akhilnarang commented on src/pages/login/login.page.jsx:
```diff
@@ -138,10 +142,19 @@ class Login extends Component {
-export default Login;
\ No newline at end of file
+const mapStateToProps = (state) => ({
+    errorMessage: state.auth.errorMessage || state.user.errorMessage, //TODO: check if this is fine
```
This is fine or not btw?

PranavBakre replied:
abtak to fine. both messages are mutually exclusive and if 1 is error, 2nd will have to be null either because it isnt called or because it gave a success for the next call to actually execute and fail
